driver/python/cursor.py

```diff
@@ -0,0 +1,216 @@
+        # Per PEP 249: The attribute is -1 in case no .execute*() has been
+        # performed on the cursor or the rowcount of the last operation
+        # cannot be determined by the interface.
+        self.rowcount = -1
+        # Per PEP 249: The arraysize attribute defaults to 1, meaning to fetch
+        # a single row at a time. However, we deviate from that, and set the
+        # default to None, allowing the backend to automatically determine the
+        # most appropriate size.
+        self.arraysize = None
+        self._query_data = None
+        self._errors = None
+
+    @property
+    def connection(self):
+        return self._connection
+
+    def _set_rowcount(self, query_results):
+        """Set the rowcount from query results.
+
+        Normally, this sets rowcount to the number of rows returned by the
+        query, but if it was a DML statement, it sets rowcount to the number
+        of modified rows.
+
+        Args:
+            query_results (Dictionary):
+                Results of a query.
+        """
+        total_rows = 0
+        num_dml_affected_rows = query_results.num_dml_affected_rows
+
+        if query_results.total_rows is not None and query_results.total_rows > 0:
+            total_rows = query_results.total_rows
+        if num_dml_affected_rows is not None and num_dml_affected_rows > 0:
+            total_rows = num_dml_affected_rows
+        self.rowcount = total_rows
+
+    def execute(self, operation, parameters=None):
+        """Prepare and execute a database operation.
+
+        Args:
+            operation (str): SFDB query string.
+
+            parameters (Union[Mapping[str, Any], Sequence[Any]]):
+                (Optional) dictionary or sequence of parameter values.
+        """
+        _LOGGER.debug('Operation:\n%s', operation)
+        self._query_data = None
+        req = api_pb2.ExecSqlRequest()
+        req.sql = operation
+        for _ in range(RETRIES):
+            try:
+                resp = self._connection.cmd_query(req)
+            except exceptions.InterfaceError as e:
+                self._errors = e.args[0]
+                _LOGGER.error('Failed to execute query:\n%s', e)
+                return False
+            _LOGGER.debug('Response:\n%s', resp)
+            if resp.status == api_pb2.ExecSqlResponse.ERROR:
+                _LOGGER.error(
+                    "Server returned an error executing:\n%s", operation)
+                self._errors = {"FAILED": operation}
+            if resp.status == api_pb2.ExecSqlResponse.REDIRECT and resp.redirect:
+                _LOGGER.debug("Redirecting to: %s", resp.redirect)
+                self._connection._redirect(resp.redirect)
+                continue
+            if resp.rows:
+                descriptors = self._get_descriptors(resp)
+                self.rowcount = self._set_data(descriptors, resp.rows)
+            else:
+                self.rowcount = 0
+            return True
```

Comment:
Maybe this loop's structure could be refactored a bit or at least the order of the two resp.status == if statements could be flipped (and maybe use else-if for the 2nd), so it is better visible that we only retry if it's a REDIRECT

Reply:
Done